### PR TITLE
add LinearQueue

### LINKED ISSUE #581: [Feature Request] Override / Disable FairQueueing via config
**Is your feature request related to a problem? Please describe.**
Nope, just an offhand request.

**Describe the solution you'd like**
Ability to switch FairQueue to a normal Queue via config.

**Describe the basic flow/steps of using this feature**
How this feature would be used:
1. Add song via play
2. Song is added to queue at the very end no matter the user who inserted the song
3. Songs play through FIFO order, regardless of identifier

**Additional context**
My friends and I really love the music bot, but we sometimes want to play a group of songs, then have another person's songs play. Having a normal queue system instead of the FairQueue system would be nice in these instances.

**Have you confirmed that this idea is not already described here?**
https://github.com/jagrosh/MusicBot/projects/1 Yessir

### PR DESCRIPTION
**Pull request by Woife5**
### This pull request...
  - [ ] Fixes a bug
  - [x] Introduces a new feature
  - [ ] Improves an existing feature
  - [ ] Boosts code quality or performance

### Description
This adds a new type of queue which maintinas the insertion order regardless of the user adding the song.
There is already a MR open about this issue #1165 but it seems to be stale.

### Purpose
Personally we have a few designated music people on our discord who add songs in their prefered order but they always get mixed up due to `FairQueue`. The newly added `LinearQueue` would circumvent this problem.

### Relevant Issue(s)
closes #581

### CODE REVIEW

File: src/main/java/com/jagrosh/jmusicbot/BotConfig.java
Comment (MichailiK) [inline src/main/java/com/jagrosh/jmusicbot/BotConfig.java:354]: Perhaps we could use an enum instead of strings?

**Comment by MichailiK** [comment]
If you'd like, [join the support server](https://discord.gg/0p9LSGoRLu6Pet0k) if you need help and/or want to discuss of the implementation for queue types. You should be able to easily find me there (Michaili#1397).

File: src/main/java/com/jagrosh/jmusicbot/audio/AudioHandler.java
Comment (MichailiK) [inline src/main/java/com/jagrosh/jmusicbot/audio/AudioHandler.java:70]: Can this be removed, or will Java think that the queue might be left unset & doesn't recognize the function call below?

File: src/main/java/com/jagrosh/jmusicbot/audio/AudioHandler.java
Comment (MichailiK) [inline src/main/java/com/jagrosh/jmusicbot/audio/AudioHandler.java:75]: I would rather name this as `setQueueType`

File: src/main/java/com/jagrosh/jmusicbot/settings/QueueType.java
Comment (MichailiK) [inline src/main/java/com/jagrosh/jmusicbot/settings/QueueType.java:26]: Seems like this random queue was unintentionally left in? There's no random queue implemented yet

**Review by MichailiK — CHANGES_REQUESTED** [review]

File: src/main/java/com/jagrosh/jmusicbot/commands/dj/QueueTypeCmd.java
Comment (MichailiK) [inline src/main/java/com/jagrosh/jmusicbot/commands/dj/QueueTypeCmd.java:62]: I suppose this is fine for now. If we are adding any more queue types, we might want a more dynamic approach.

File: src/main/java/com/jagrosh/jmusicbot/settings/SettingsManager.java
Comment (MichailiK) [inline src/main/java/com/jagrosh/jmusicbot/settings/SettingsManager.java:34]: Having this constant here is actually nice & fine I guess, albeit out of scope of this PR.

File: src/main/java/com/jagrosh/jmusicbot/audio/AudioHandler.java
Comment (Woife5) [reply src/main/java/com/jagrosh/jmusicbot/audio/AudioHandler.java:70]: The method `setQueueType()` requires the queue to already be initialized, since it calls the "copy constructors" which copy the internal queue of the currently active queue.
I do not know a more elegant solution off the top of my head, any ideas?

Comment (Woife5) [reply src/main/java/com/jagrosh/jmusicbot/audio/AudioHandler.java:70]: Adding a null-check to the `AbstractQueue` constructor works, I will do that for now

**Comment by MichailiK** [comment]
As for 1. I'm not sure if I was clear, but the diff for `QueueType` should be something like this:

```diff
-   LINEAR("\u23E9", "Linear"),
+   LINEAR("\u23E9", "Linear", LinearQueue::new),
-   FAIR("\uD83D\uDD22", "Fair");
+   FAIR("\uD83D\uDD22", "Fair", FairQueue::new);

    private final String userFriendlyName;
    private final String emoji;
+   private final Function<AbstractQueue<?>, ? extends AbstractQueue<?>> supplier;

-   QueueType(final String emoji, final String userFriendlyName) {
+   QueueType(final String emoji, final String userFriendlyName, Function<AbstractQueue<?>, ? extends AbstractQueue<?>> supplier) {
        this.userFriendlyName = userFriendlyName;
        this.emoji = emoji;
+       this.supplier = supplier;
    }

+   public AbstractQueue<?> createInstance(AbstractQueue<?> previousInstance)
+   {
+       return supplier.apply(previousInstance);
+   }
```

**Comment by Sanduhr32** [comment]
Functions are for mapping the first generic type argument to the second generic type. Supplier is probably what you want

**Comment by Woife5** [comment]
I think a `Supplier` would not work with the constructor argument `previousInstance`, right? Because it takes no arguments

**Comment by Sanduhr32** [comment]
yes, but why do you need the argument, the function does absolutely nothing with it.

File: src/main/java/com/jagrosh/jmusicbot/audio/AudioHandler.java
Comment (Woife5) [inline src/main/java/com/jagrosh/jmusicbot/audio/AudioHandler.java:74]: @MichailiK I currently cannot think of any way to get rid of this cast

Comment (MichailiK) [reply src/main/java/com/jagrosh/jmusicbot/audio/AudioHandler.java:74]: Hm, I believe that we will need to create our own functional interface. We can make a `QueueSupplier` in the utils package:
```java
@FunctionalInterface
public interface QueueSupplier
{
    <T extends Queueable> AbstractQueue<T> apply(AbstractQueue<T> previous);
}
```

Then, we can go in our queuetype and replace all the `Function<...>`s with `QueueSupplier`. Then you can finally remove this cast.

Comment (Woife5) [reply src/main/java/com/jagrosh/jmusicbot/audio/AudioHandler.java:74]: That seems to have done the trick, thanks ^^

File: src/main/java/com/jagrosh/jmusicbot/BotConfig.java
Comment (MichailiK) [inline src/main/java/com/jagrosh/jmusicbot/BotConfig.java:19]: Unnecessary diff

File: src/main/java/com/jagrosh/jmusicbot/BotConfig.java
Comment (MichailiK) [inline src/main/java/com/jagrosh/jmusicbot/BotConfig.java:353]: Unnecessary diff

File: src/main/java/com/jagrosh/jmusicbot/commands/dj/QueueTypeCmd.java
Comment (MichailiK) [inline src/main/java/com/jagrosh/jmusicbot/commands/dj/QueueTypeCmd.java:2]: Might want to update the copyright year to 2022

File: src/main/java/com/jagrosh/jmusicbot/commands/dj/QueueTypeCmd.java
Comment (MichailiK) [inline src/main/java/com/jagrosh/jmusicbot/commands/dj/QueueTypeCmd.java:51]: Seeing as the volume command puts a volume emoji (based on the current volume) in the beginning, I would say you should put the emoji at the beginning as well.

File: src/main/java/com/jagrosh/jmusicbot/commands/dj/QueueTypeCmd.java
Comment (MichailiK) [inline src/main/java/com/jagrosh/jmusicbot/commands/dj/QueueTypeCmd.java:77]: I would also add the queue's emoji to the beginning here.

File: src/main/java/com/jagrosh/jmusicbot/commands/dj/QueueTypeCmd.java
Comment (MichailiK) [inline src/main/java/com/jagrosh/jmusicbot/commands/dj/QueueTypeCmd.java:69]: Seeing as a "is this already set" check doesn't exist with the other settings-related commands, I would remove this to stay consistent.

File: src/main/java/com/jagrosh/jmusicbot/queue/AbstractQueue.java
Comment (MichailiK) [inline src/main/java/com/jagrosh/jmusicbot/queue/AbstractQueue.java:2]: Change the copyright year to 2022 here as well

File: src/main/java/com/jagrosh/jmusicbot/queue/LinearQueue.java
Comment (MichailiK) [inline src/main/java/com/jagrosh/jmusicbot/queue/LinearQueue.java:2]: Change the copyright year to 2022 here as well

File: src/main/java/com/jagrosh/jmusicbot/settings/QueueType.java
Comment (MichailiK) [inline src/main/java/com/jagrosh/jmusicbot/settings/QueueType.java:2]: Change the copyright year to 2022 here as well

File: src/main/java/com/jagrosh/jmusicbot/utils/QueueSupplier.java
Comment (MichailiK) [inline src/main/java/com/jagrosh/jmusicbot/utils/QueueSupplier.java:2]: Change the copyright year to 2022 here as well

**Review by MichailiK — CHANGES_REQUESTED** [review]
This looks pretty great now! There are just these nitpicks I have about this PR. Once these are resolved, this PR will reach *perfection*

**Comment by MichailiK** [comment]
Oh yeah, I almost forgot: You should add an entry to the `SettingsCmd`. Something like this below Repeat Mode should suffice:

```java
+ "\nQueue Type: " (s.getQueueType() == QueueType.FAIR
                     ? s.getQueueType().getUserFriendlyName() 
                     : "**"+s.getQueueType().getUserFriendlyName()+"**" )
```

(The fair queue check is there, in case it got changed from the default value. Settings that are changed from default are bold too.)

Once the queue type is being displayed in the settings, this should be good to merge!

**Comment by Woife5** [comment]
One thing I just thought of: I added `QueueTypeCmd` as DJ command, is this okay or should it be an Admin command? @MichailiK

**Comment by MichailiK** [comment]
Having this as a DJ command seems the most ideal to me for now. jagrosh will take a look at the PR & optionally give his thoughts.

File: src/main/java/com/jagrosh/jmusicbot/queue/AbstractQueue.java
Comment (MichailiK) [inline src/main/java/com/jagrosh/jmusicbot/queue/AbstractQueue.java:29]: I would add a `protected` modifier here.

File: src/main/java/com/jagrosh/jmusicbot/queue/AbstractQueue.java
Comment (MichailiK) [inline src/main/java/com/jagrosh/jmusicbot/queue/AbstractQueue.java:37]: I would also remove this constructor here, since we won't be using it anymore. Remove it from `FairQueue` & `LinearQueue`

File: src/main/java/com/jagrosh/jmusicbot/queue/FairQueue.java
Comment (MichailiK) [inline src/main/java/com/jagrosh/jmusicbot/queue/FairQueue.java:30]: Put the curly braces here in a new line.
```suggestion
public class FairQueue<T extends Queueable> extends AbstractQueue<T>
{

    public FairQueue(AbstractQueue<T> queue)
    {
        super(queue);
    }
```

File: src/main/java/com/jagrosh/jmusicbot/settings/QueueType.java
Comment (MichailiK) [inline src/main/java/com/jagrosh/jmusicbot/settings/QueueType.java:64]: Remove the unnecessary type parameter in the constructor and put all the curly braces here in a new line.
```suggestion
    QueueType(final String emoji, final String userFriendlyName, QueueSupplier supplier)
    {
        this.userFriendlyName = userFriendlyName;
        this.emoji = emoji;
        this.supplier = supplier;
    }

    public static List<String> getNames()
    {
        return Arrays.stream(QueueType.values())
                .map(type -> type.name().toLowerCase())
                .collect(Collectors.toList());
    }

    public <T extends Queueable> AbstractQueue<T> createInstance(AbstractQueue<T> previous)
    {
        return supplier.apply(previous);
    }

    public String getUserFriendlyName()
    {
        return userFriendlyName;
    }

    public String getEmoji()
    {
        return emoji;
    }
```

File: src/main/java/com/jagrosh/jmusicbot/queue/LinearQueue.java
Comment (MichailiK) [inline src/main/java/com/jagrosh/jmusicbot/queue/LinearQueue.java:25]: Put the curly braces in a new line.
```suggestion
    public LinearQueue(AbstractQueue<T> queue)
    {
```

**Review by MichailiK — CHANGES_REQUESTED** [review]
Hellooo! I just noticed i missed a few things that should change before this PR is perfect

**Review by MichailiK — APPROVED** [review]
Looks good! @jagrosh should take a look & merge if all's good.

**Comment by LinearQueueIsGood** [comment]
I would really like a linear queue feature with this please consider adding it thank you.

**Comment by JiimJaam** [comment]
Is there way to just add this to my own hosted version of the bot temporarily until it's officially merged (if that even happens)?

**Comment by Woife5** [comment]
> Is there way to just add this to my own hosted version of the bot temporarily until it's officially merged (if that even happens)?

You can basically just build the bot yourself with this feature included. If you need any help with that you can message me on Discord: Wolfgang#3620

**Review by MichailiK — APPROVED** [review]

File: src/main/java/com/jagrosh/jmusicbot/settings/SettingsManager.java
Comment (jagrosh) [inline src/main/java/com/jagrosh/jmusicbot/settings/SettingsManager.java:36]: This isn't needed anymore because of a different merged PR

File: src/main/java/com/jagrosh/jmusicbot/commands/dj/QueueTypeCmd.java
Comment (jagrosh) [inline src/main/java/com/jagrosh/jmusicbot/commands/dj/QueueTypeCmd.java:1]: QueueTypeCmd feels more like an admin command (modifying a server-wide setting) rather than a dj command (modifying the existing playback)

File: src/main/java/com/jagrosh/jmusicbot/commands/dj/QueueTypeCmd.java
Comment (jagrosh) [inline src/main/java/com/jagrosh/jmusicbot/commands/dj/QueueTypeCmd.java:79]: This also wouldn't be necessary as an admin command

File: src/main/java/com/jagrosh/jmusicbot/utils/QueueSupplier.java
Comment (jagrosh) [inline src/main/java/com/jagrosh/jmusicbot/utils/QueueSupplier.java:1]: I think this file could be moved to the queue folder, since it's only ever used by queue-related things

**Review by jagrosh — CHANGES_REQUESTED** [review]

File: src/main/java/com/jagrosh/jmusicbot/commands/dj/QueueTypeCmd.java
Comment (Woife5) [reply src/main/java/com/jagrosh/jmusicbot/commands/dj/QueueTypeCmd.java]: makes sense, i moved it 👾

File: src/main/java/com/jagrosh/jmusicbot/utils/QueueSupplier.java
Comment (Woife5) [reply src/main/java/com/jagrosh/jmusicbot/utils/QueueSupplier.java]: done

**Review by MichailiK — APPROVED** [review]

**Pull request merged**
